### ISSUE TITLE
Intellisense for scripts created by user

### ISSUE THREAD

**Issue by gmastergreatee** [loganch/AutoIt-VSCode#14]
I don't know if this can be already achieved via tweaking the settings

While including custom scripts(imported script is in the same directory as the one being being imported from), also add intelliSense for methods in the imported script.

I tried 

```
#include <NomadMemory.au3>

```

script available [here](http://forum.cheatengine.org/viewtopic.php?t=517600)
still no intellisense for functions like _MemoryOpen, _MemoryWrite, etc...

Thank you for your time
Feel free to close.

**Comment by Worlor** [comment]
Oh yes it would be a nice feature

**Comment by loganch** [comment]
Believe me, I've been brainstorming since I had the signature helpers working on how to incorporate functions from custom scripts.

**Comment by Worlor** [comment]
What is blocking you ?

**Comment by Worlor** [comment]
I tried the functionnality on the last commit, it seems to work great !
Do you think it's possible to add function parameters in intellisense ?

**Comment by loganch** [comment]
What was blocking me was figuring out just how to structure the code to get the functions. 
As for the parameters, do you mean the signature help that pops up as you advance through a function?

**Comment by Worlor** [comment]
Yes, like this : 
![image](https://user-images.githubusercontent.com/3814031/28772068-539e2bd0-75e5-11e7-864b-19b8f0b2f8de.png)

But without the doc ;)

**Comment by R-THopper** [comment]
managing autocomplete of included user files would also be extremely helpful. if possible :)

**Comment by loganch** [comment]
I've started a branch with a start on SignathureHelp Intellisense for user-included functions. Hopefully, I can get it up to speed and released soon-ish.

**Comment by R-THopper** [comment]
are you maintaining the extension on the vscode store? if so, i will take a copy of said branch and i have a couple ideas on intellisence for autoitobject classes. ill let you know if it ever ends up somewhere. Thanks for your work on this btw, i use autoit alot... ALOT...... and this is speeding up developpment on automated testing so much,

**Comment by loganch** [comment]
Yes, I took over development of it a year ago and made this the main repo.

**Comment by R-THopper** [comment]
ok! good to know. So ill let you know about autoit object integration. it would take autoit development to the next level to have intellisence on objects. if you already have a vision on the subject, let me know. otherwise ill stay updated on the branch you made and will let you know when i have something that works.

**Comment by loganch** [comment]
I haven't encountered AutoIt Objects yet, but I've found myself wishing for some OO features while working on scripts. Do you have any links to information about it? I'm glad the extension has been of use to you.

**Comment by Worlor** [comment]
For OOP there is an extension, but not very easy to use : 
[https://www.autoitscript.com/forum/topic/178542-oop-extender-v03-stable-real-object-oriented-programming-with-autoit/](https://www.autoitscript.com/forum/topic/178542-oop-extender-v03-stable-real-object-oriented-programming-with-autoit/)

**Comment by R-THopper** [comment]
this is way more mature (you can use file includes and class inheritance, it compiles well and life in general is simpler)
https://www.autoitscript.com/forum/topic/110379-autoitobject-udf/
this udf is what i am working to get intellisense on. basically all i am doing is mapping addMethods and addProperty values, but i will get the parameters of the function to show in tooltip somehow.. thats where i am atm

**Comment by loganch** [comment]
On the userSigHelp branch I've enabled sig help for functions declared inside the current document, if you'd like to give a try before I make the release.

**Comment by Worlor** [comment]
I am trying the new version right now (on branch usersighelp). 

It works pretty well and help to call UDF.

However, It take some time (more than with other files) to have the autocomplete with this UDF : [Lib 3270](https://github.com/Worlor/AutoIt-3270)

**Comment by buxters** [comment]
Hello! Are you working on this direction?
I really look forward to this functionality :)

**Comment by loganch** [comment]
Hi @buxters, both Completions and Signature helpers for individual user scripts have been implemented for a few versions now. 

I figure this Issue not being closed makes it seems as though I may not have finished it. 

Could you verify that it's working properly for you and I'll go ahead and close this out?

**Comment by buxters** [comment]
Hi @loganch, I tried: the tooltip is displayed, but there is no description of the function and parameters.
![image](https://user-images.githubusercontent.com/6255003/43013993-c806aa96-8c5b-11e8-8659-69f259c2a490.png)
My settings:
![image](https://user-images.githubusercontent.com/6255003/43014023-e051dcc4-8c5b-11e8-836a-12d8a4e8ec2b.png)
function description in udf:
![image](https://user-images.githubusercontent.com/6255003/43014064-069e534e-8c5c-11e8-8511-3376c0b20284.png)

Maybe I misconfigured something :(((

**Comment by loganch** [comment]
Ah, I see what you mean now. 
The extension is working as designed, I just haven't implemented descriptions (yet?) as there's no standard way that they are written into the code.

**Comment by buxters** [comment]
@loganch, did you plan to implement this function?
thank

**Comment by loganch** [comment]
@buxters I think I can eventually pull that information when creating the signature helpers, but I'll have to figure out how to make it work without a standard layout (not sure I have enough pull to set a standard, lol).

**Comment by buxters** [comment]
Maybe, at the first, implement on the basis of a template :) ?
For example, based on the recommended template on the site - [UDF-spec](https://www.autoitscript.com/wiki/UDF-spec)

```
; #FUNCTION# =====
; Name...........: 
; Description ...: 
; Syntax.........: 
; Parameters ....: 
; Return values .: 
; Author ........: 
; Modified.......:
; Remarks .......: 
; Related .......: 
; Link ..........: 
; Example .......: 
==================
```
